### PR TITLE
redo kubernetes DNS setup to fix latency problems

### PR DESCRIPTION
**Pull request by szuecs**
* switch dns resolver to coredns, which has better metrics support
* enable node local dnsmasq cache to scale out dns
* node creation: lookup node ip to create a node local resolv.conf target to hit the daemonset dnsmasq cache before falling back to coredns service ip
* adjust node-local svc for dnsmasq monitoring
* coredns set resolve: pods disabled

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

### CODE REVIEW

**Comment by szuecs** [comment]
We might need to delete the kube-dns deployment and service to make this rolling update work correctly.

**Comment by hjacobs** [comment]
@szuecs any benchmark results for this change (yet)?

**Comment by szuecs** [comment]
@hjacobs yes. No matter what I did dnsblast with 200req/s destroyed the dns setup:

vertical or horizontal scale or use more cache or use --no-negcache or a kube-dns replaced by coredns, I got 5s timeouts:

image

New setup looks much better with 100ms:
image
New setup is coredns resolver and dnsmasq local dns cache with more cache size and --no-ngecache. All pods lookup first from local dnsmasq with caching.

File: cluster/manifests/kube-dns/node-local-daemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        volumeMounts:
+        - name: kube-dns-config
+          mountPath: /etc/k8s/dns/dnsmasq-nanny
+      - name: sidecar
+        image: registry.opensource.zalan.do/teapot/k8s-dns-sidecar-amd64:1.14.10
+        securityContext:
+          privileged: true
+        livenessProbe:
+          httpGet:
+            path: /metrics
```
Comment (mikkeloscar) [inline cluster/manifests/kube-dns/node-local-daemonset.yaml:78]: @szuecs Should we change this to not call the metrics endpoint for liveness? You mentioned it's a lot of data for a liveness probe :)

File: cluster/manifests/kube-dns/node-local-daemonset.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        - containerPort: 53
+          name: dns
+          protocol: UDP
+        - containerPort: 53
+          name: dns-tcp
+          protocol: TCP
+        # see: https://github.com/kubernetes/kubernetes/issues/29055 for details
+        resources:
+          requests:
+            cpu: 150m
```
Comment (mikkeloscar) [inline cluster/manifests/kube-dns/node-local-daemonset.yaml]: @szuecs do you have any metrics on how much CPU is needed? It seems like a lot to use `150m` on each node for part of the base infrastructure.

(we should see how much we use for all daemonsets combined).

File: cluster/node-pools/master-default/userdata.clc.yaml [hunk 3]
```diff
@@ -150,7 +150,7 @@ systemd:
       --node-labels=kubernetes.io/node-pool={{ .NodePool.Name }} \
       --node-labels={{ .Values.node_labels }} \
       --pod-manifest-path=/etc/kubernetes/manifests \
-      --cluster-dns=10.3.0.10,10.3.0.11 \
+      --cluster-dns=127.0.0.1:30053,10.3.0.10,10.3.0.11 \
```
Comment (mikkeloscar) [inline cluster/node-pools/master-default/userdata.clc.yaml]: I guess this is just left by mistake.

**Comment by mikkeloscar** [comment]
We probably need to split this up and roll it out in two steps. First rolling out the daemonset and then rolling out the nodes where kubelet is configured to use the node ip in `resolv.conf`.

**Comment by szuecs** [comment]
I get this log:

```time="2018-06-01T17:28:43Z" level=fatal msg="Fail to provision: run kubectl failed: exit status 1: The Service "coredns" is invalid: spec.clusterIP: Invalid value: "10.3.0.10": provided IP is already allocated```

@mikkeloscar do you know how to fix it and where?

File: cluster/manifests/kube-dns/node-local-daemonset.yaml (hunk 1 above)
Comment (szuecs) [reply cluster/manifests/kube-dns/node-local-daemonset.yaml:78]: I would go for the defaults first

File: cluster/manifests/kube-dns/node-local-daemonset.yaml (hunk 2 above)
Comment (szuecs) [reply cluster/manifests/kube-dns/node-local-daemonset.yaml]: I will test with less requests, I think it's less than 50m.

**Comment by szuecs** [comment]
step 1: https://github.com/zalando-incubator/kubernetes-on-aws/pull/1112

**Pull request closed without merging**